fuse/src/main/java/tachyon/fuse/TachyonFuseFs.java

```diff
@@ -0,0 +1,671 @@
+   */
+  @Override
+  public String getFSName() {
+    return "tachyon-fuse";
+  }
+
+  /**
+   * Create a new dir.
+   * @param path the path on the FS of the new dir
+   * @param mode Dir creation flags (IGNORED)
+   * @return 0 on success, a negative value on error
+   */
+  @Override
+  public int mkdir(String path, @mode_t long mode) {
+    final TachyonURI turi = mPathResolverCache.getUnchecked(path);
+    LOG.trace("mkdir({}) [Tachyon: {}]", path, turi);
+    try {
+      mTFS.mkdir(turi);
+    } catch (FileAlreadyExistsException e) {
+      LOG.debug("Cannot make dir. {} already exists", path, e);
+      return -ErrorCodes.EEXIST();
+    } catch (InvalidPathException e) {
+      LOG.debug("Cannot make dir. Invalid path: {}", path, e);
+      return -ErrorCodes.ENOENT();
+    } catch (IOException e) {
+      LOG.error("Cannot make dir. IOException: {}", path, e);
+      return -ErrorCodes.EIO();
+    } catch (TachyonException e) {
+      LOG.error("Cannot make dir. {}", path, e);
+      return -ErrorCodes.EFAULT();
+    } catch (Throwable e) {
+      LOG.error("Unexpected exception on {}", path, e);
+      return -ErrorCodes.EFAULT();
+    }
+
+    return 0;
+  }
+
+  /**
+   * Open an existing file for reading.
+   *
+   * Note that the open mode <emph>must</emph> be
+   * O_RDONLY, otherwise the open will fail. This is due to
+   * the Tachyon "write-once/read-many-times" file model.
+   *
+   * @param path the FS path of the file to open
+   * @param fi FileInfo data structure kept by FUSE
+   * @return 0 on success, a negative value on error
+   */
+  @Override
+  public int open(String path, FuseFileInfo fi) {
+    final TachyonURI turi = mPathResolverCache.getUnchecked(path);
+    final int flags = fi.flags.get();
+    LOG.trace("open({}, 0x{}) [Tachyon: {}]", path, Integer.toHexString(flags), turi);
+
+    if ((flags & 3) != O_RDONLY.intValue()) {
+      LOG.error("Files can only be opened in O_RDONLY mode ({})", path);
+      return -ErrorCodes.EACCES();
+    }
+    try {
+      final TachyonFile tf = mTFS.openIfExists(turi);
+      if (tf == null) {
+        LOG.error("File {} does not exist", turi);
+        return -ErrorCodes.ENOENT();
+      }
+      final FileInfo tfi = mTFS.getInfo(tf);
+      if (tfi.isFolder) {
+        LOG.error("File {} is a directory", turi);
+        return -ErrorCodes.EISDIR();
+      }
+      final OpenFileEntry ofe =
+          new OpenFileEntry(mTFS.getInStream(tf), null);
+
+      synchronized (mOpenFiles) {
+        if (mOpenFiles.size() == MAX_OPEN_FILES) {
+          LOG.error("Cannot open {}: too many open files", turi);
+          return ErrorCodes.EMFILE();
+        }
+        mOpenFiles.put(mNextOpenFileId, ofe);
+        fi.fh.set(mNextOpenFileId);
+
+        // Assuming I will never wrap around (2^64 open files are quite a lot anyway)
+        mNextOpenFileId += 1;
+      }
+
+    } catch (FileDoesNotExistException e) {
+      LOG.debug("File does not exist {}", path, e);
```

Comment:
I would expect them quite regularly. For example imagine the case:
`$ cat misspelledFileName.txt`